File: src/main/java/com/tuhuynh/jerrymouse/core/nio/AsyncHelper.java
package com.tuhuynh.jerrymouse.core.nio;

import com.tuhuynh.jerrymouse.core.RequestBinder.HttpResponse;
import lombok.Getter;

import java.util.concurrent.CompletableFuture;
import java.util.function.Consumer;

@Getter
public final class AsyncHelper {
    CompletableFuture<HttpResponse> promise;

    private AsyncHelper(final CompletableFuture<HttpResponse> promise) {
        this.promise = promise;
    }

    public static AsyncHelper make() {
        CompletableFuture<HttpResponse> completableFuture = new CompletableFuture<>();
        return new AsyncHelper(completableFuture);
    }

    public <T> void resolve(final T t) {
        promise.complete(HttpResponse.of(t));
    }

    public <T extends Error> void reject(final T t) {
        promise.completeExceptionally(t);
    }

    public CompletableFuture<Void> then(Consumer<? super HttpResponse> action) {
        return promise.thenAccept(action);
    }

    public CompletableFuture<HttpResponse> submit() {
        return promise;
    }
}

File: src/main/scala/com/tuhuynh/jerrymouse/core/nio/AsyncHelper.scala
package com.tuhuynh.jerrymouse.core.nio

import java.util.concurrent.CompletableFuture
import java.util.function.Consumer

import com.tuhuynh.jerrymouse.core.RequestBinderBase
import com.tuhuynh.jerrymouse.core.RequestBinderBase.HttpResponse

class AsyncHelper(val promise: CompletableFuture[HttpResponse] = new CompletableFuture[RequestBinderBase.HttpResponse]) {
  def resolve[T](t: T): Unit = {
    promise.complete(HttpResponse.of(t))
  }

  def reject[T <: Error](t: T): Unit = {
    promise.complete(HttpResponse.of(t))
  }

  def then(action: Consumer[_ >: RequestBinderBase.HttpResponse]): CompletableFuture[Void] = promise.thenAccept(action)

  def submit(): CompletableFuture[HttpResponse] = promise
}
